// build/webpack.config.dev.js
var path = require('path');
const config = require('config');
const HappyPack = require('happypack');
const webpack = require('webpack');
module.exports = {
    entry: './src/index.js',
    output: {
        filename: 'index.js',
        chunkFilename: '[id].chunk.js',
        path: path.resolve(config.root, 'dist'),
        publicPath: '/js/'
    },
    devtool: 'inline-source-map',
    cache: true,
    module: {
        rules: [{
                test: /\.less$/,
                exclude: [
                    /\.use(able)?\.less$/
                ],
                loaders: [
                    'style-loader',
                    'css-loader',
                    'px2rem-loader?remUnit=37.5',
                    'less-loader'
                ]
            },
            {
                test: /\.use(able)?\.less$/,
                exclude: /node_modules/,
                use: [
                    'style-loader/useable',
                    'css-loader',
                    'px2rem-loader?remUnit=37.5',
                    'less-loader'
                ]
            }, {
                test: /\.(png|svg|jpg|gif)/,
                use: [
                    'file-loader'
                ]
            }, {
                test: /\.js$/,
                exclude: /node_modules/,
                loader: 'happypack/loader?id=js',
            }
        ]
    },
    plugins: [
        new webpack.DllReferencePlugin({
            // context: path.resolve(config.path.dll),
            manifest: require(path.resolve(config.path.dll, "vendor-manifest.json")) 
        }),
        new HappyPack({
            id: 'js',
            threads: 4,
            // debug: true,
            loaders: [ 'babel-loader' ]
        })
    ],
    resolve: {
        //配置别名，在项目中可缩减引用路径
        alias: {
            common: path.resolve(config.root, 'common'),
            src: path.resolve(config.root, 'src'),
            lib: path.resolve(config.root, 'src/lib'),
            images: path.resolve(config.root, 'src/images')
        }
    }
};
